src/storage/checkpoint/table_data_writer.cpp

```diff
@@ -66,13 +66,21 @@ void SingleFileTableDataWriter::FinalizeTable(TableStatistics &&global_stats, Da
 		row_group_serializer.End();
 	}
 
-	auto index_pointers = info->indexes.SerializeIndexes(table_data_writer);
-
 	// Now begin the metadata as a unit
 	// Pointer to the table itself goes to the metadata stream.
-	metadata_serializer.WriteProperty(101, "table_pointer", pointer);
-	metadata_serializer.WriteProperty(102, "total_rows", total_rows);
-	metadata_serializer.WriteProperty(103, "index_pointers", index_pointers);
+	serializer.WriteProperty(101, "table_pointer", pointer);
+	serializer.WriteProperty(102, "total_rows", total_rows);
+
+	// STABLE STORAGE NOTE: We write an index pointers vector with exactly one invalid index pointer with a valid
+	// offset, because when reading the file, we don't know if we're reading an older storage version (with
+	// BlockPointers) or not. The combination of invalid block ID and valid offset allows us to distinguish
+	vector<BlockPointer> index_pointers;
+	index_pointers.emplace_back(INVALID_BLOCK, 42);
+	serializer.WriteProperty(103, "index_pointers", index_pointers);
+
+	auto index_storage_infos = info->indexes.GetStorageInfos();
+	serializer.WriteList(104, "index_storage_infos", index_storage_infos.size(),
+	                     [&](Serializer::List &list, idx_t i) { list.WriteElement(index_storage_infos[i]); });
```

Comment:
move it to a modular function as well?

These stable storage codes are small but also look a bit dirty, maybe they should be moved somewhere (static functions in a Index_storage with the comments?